### PR TITLE
Feature/improve mfa

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Define a default MFA method in Akamai MFA configuration.
Support for hardware token like yubikey (TOTP)

# Impacts
RADIUS cli/vpn

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries

## Enhancements
* Support for hardware token like yubikey for Akamai MFA
* Added sms/phone call as default method in configuration

### CODE REVIEW

File: html/pfappserver/lib/pfappserver/Form/Config/Mfa/Akamai.pm [hunk 1]
```diff
@@ -64,6 +64,8 @@ has_field 'radius_mfa_method' =>
    options =>
    [
     { value => 'push', label => 'Push' },
+    { value => 'sms' , label => 'Sms' },
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/Mfa/Akamai.pm]: if possible, we should use "SMS" instead of "Sms"

File: lib/pf/mfa/Akamai.pm [hunk 2]
```diff
@@ -151,32 +152,43 @@ sub check_user {
             return $ACTIONS{'push'}->($self,$default_device[0]->{'device'},$username);
        }
     }
-    elsif ($self->radius_mfa_method eq 'strip-otp' || $self->radius_mfa_method eq 'second-password') {
+    elsif ($self->radius_mfa_method eq 'strip-otp' || $self->radius_mfa_method eq 'second-password' || $self->radius_mfa_method eq 'sms' || $self->radius_mfa_method eq 'phone') {
```
Comment (julsemaan) [inline lib/pf/mfa/Akamai.pm]: I think that this `if` is starting to get a bit too fat (insert pun about you know who here)

A lookup in a map would make things simpler/clearer

File: lib/pf/mfa/Akamai.pm [hunk 3]
```diff
@@ -151,32 +152,43 @@ sub check_user {
...
+                        return $ACTIONS{$method}->($self,$device->{'device'},$username,$1,$devices);
                     } else {
-                        $logger->warn("Unsuported method on device ".$device->{'name'});
+                        $logger->info("Unsuported method on device ".$device->{'name'});
                         return $FALSE;
                     }
                 }
             } else {
-                $logger->warn("Method not supported");
+                $logger->info("Method not supported");
```
Comment (julsemaan) [inline lib/pf/mfa/Akamai.pm:183]: Wouldn't that be more of an error? Or can this happen under normal circumstances?

File: lib/pf/mfa/Akamai.pm [hunk 4]
```diff
@@ -151,32 +152,43 @@ sub check_user {
...
             } elsif ($otp =~ /^(sms|push|phone)(\d?)$/i) {
                 my @device = $self->select_phone($devices->{'result'}->{'devices'}, $2);
                 my $method = $1;
                 foreach my $device (@device) {
                     if ( grep $_ =~ $METHOD_ALIAS{$method}, @{$device->{'methods'}}) {
-                        return $ACTIONS{$method}->($self,$device->{'device'},$username,$1);
+                        return $ACTIONS{$method}->($self,$device->{'device'},$username,$1,$devices);
                     } else {
-                        $logger->warn("Unsuported method on device ".$device->{'name'});
+                        $logger->info("Unsuported method on device ".$device->{'name'});
```
Comment (julsemaan) [inline lib/pf/mfa/Akamai.pm]: Wouldn't that be more of an error? Or can this happen under normal circumstances?

**Comment by julsemaan** [comment]
I don't think this should be a +1 though, this should be in 12.1 as it is more a feature than a bug fix

File: lib/pf/mfa/Akamai.pm (hunk 3 above)
Comment (fdurand) [reply lib/pf/mfa/Akamai.pm:183]: Yes it´s the feature of the device that providing the method supported. So if you are doing push on a device that doesn´t support it then it´s more a notice than an error.

**Review by julsemaan — APPROVED** [review]

**Comment by julsemaan** [comment]
I just approved this PR

@jrouzierinverse, your review was requested too, can you review this and merge

File: lib/pf/mfa/Akamai.pm (hunk 4 above)
Comment (jrouzierinverse) [reply lib/pf/mfa/Akamai.pm]: ```Unsuported => Unsupported ```

File: lib/pf/mfa/Akamai.pm [hunk 5]
```diff
@@ -151,32 +152,43 @@ sub check_user {
...
-    elsif ($self->radius_mfa_method eq 'strip-otp' || $self->radius_mfa_method eq 'second-password') {
+    elsif ($self->radius_mfa_method eq 'strip-otp' || $self->radius_mfa_method eq 'second-password' || $self->radius_mfa_method eq 'sms' || $self->radius_mfa_method eq 'phone') {
         if (defined $otp) {
             if ($otp =~ /^\d{6,6}$/ || $otp =~ /^\d{16,16}$/) {
                 if ( grep $_ eq 'totp', @{$default_device[0]->{'methods'}}) {
-                    return $ACTIONS{'totp'}->($self,$default_device[0]->{'device'},$username,$otp);
+                    return $ACTIONS{'totp'}->($self,$default_device[0]->{'device'},$username,$otp,$devices);
                 } else {
-                    $logger->warn("Unsuported method totp on device ".$default_device[0]->{'name'});
+                    $logger->info("Unsuported method totp on device ".$default_device[0]->{'name'});
```
Comment (jrouzierinverse) [inline lib/pf/mfa/Akamai.pm]: Typo ```Unsuported => Unsupported```

File: lib/pf/mfa/Akamai.pm [hunk 6]
```diff
@@ -151,32 +152,43 @@ sub check_user {
...
+                        $logger->info("Unsuported method on device ".$device->{'name'});
                         return $FALSE;
                     }
                 }
             } else {
-                $logger->warn("Method not supported");
+                $logger->info("Method not supported");
                 return $FALSE;
             }
+        } elsif ($self->radius_mfa_method eq 'sms' || $self->radius_mfa_method eq 'phone') {
```
Comment (jrouzierinverse) [inline lib/pf/mfa/Akamai.pm:186]: This looks like dead code, as it will never be called because SMS and phone are included above.

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

File: lib/pf/mfa/Akamai.pm (hunk 6 above)
Comment (jrouzierinverse) [reply lib/pf/mfa/Akamai.pm:186]: Ignore

File: lib/pf/mfa/Akamai.pm (hunk 2 above)
Comment (jrouzierinverse) [reply lib/pf/mfa/Akamai.pm]: It looks like just an ```else``` would do since that is the list of all the other choices.

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

**Comment by fdurand** [comment]
Change done for PR

File: lib/pf/mfa/Akamai.pm [hunk 7]
```diff
@@ -189,7 +189,7 @@ sub check_user {
                 if ( grep $_ =~ $METHOD_ALIAS{$self->radius_mfa_method}, @{$device->{'methods'}}) {
                     return $ACTIONS{$self->radius_mfa_method}->($self,$device->{'device'},$username,$self->radius_mfa_method);
                 } else {
-                    $logger->info("Unsuported method on device ".$device->{'name'});
+                    $logger->info("Unsup orted method on device ".$device->{'name'});
```
Comment (jrouzierinverse) [inline lib/pf/mfa/Akamai.pm]: "Unsup orted" -> "Unsupported"

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

**Comment by fdurand** [comment]
BUMP !

File: lib/pf/mfa/Akamai.pm [hunk 8]
```diff
@@ -133,12 +133,17 @@ sub check_user {
        return $FALSE;
     }
     if (exists($devices->{'result'}->{'policy_decision'})) {
+        if ($devices->{'result'}->{'policy_decision'} eq "bypass") {
+            $logger->info("Policy descision is bypass, allow access");
```
Comment (jrouzierinverse) [inline lib/pf/mfa/Akamai.pm]: Typo ```descision -> decision```

File: lib/pf/mfa/Akamai.pm [hunk 9]
```diff
@@ -151,32 +156,43 @@ sub check_user {
...
             } elsif ($otp =~ /^(sms|push|phone)(\d?)$/i) {
                 my @device = $self->select_phone($devices->{'result'}->{'devices'}, $2);
                 my $method = $1;
                 foreach my $device (@device) {
                     if ( grep $_ =~ $METHOD_ALIAS{$method}, @{$device->{'methods'}}) {
-                        return $ACTIONS{$method}->($self,$device->{'device'},$username,$1);
+                        return $ACTIONS{$method}->($self,$device->{'device'},$username,$1,$devices);
                     } else {
-                        $logger->warn("Unsuported method on device ".$device->{'name'});
+                        $logger->info("Unsupported method on device ".$device->{'name'});
```
Comment (jrouzierinverse) [inline lib/pf/mfa/Akamai.pm:178]: Typo ```Unsuported => Unsupported```

File: lib/pf/mfa/Akamai.pm [hunk 10]
```diff
@@ -151,32 +156,43 @@ sub check_user {
...
+                $logger->info("Method not supported");
                 return $FALSE;
             }
+        } elsif ($self->radius_mfa_method eq 'sms' || $self->radius_mfa_method eq 'phone') {
+            my @device = $self->select_phone($devices->{'result'}->{'devices'}, undef);
+            foreach my $device (@device) {
+                if ( grep $_ =~ $METHOD_ALIAS{$self->radius_mfa_method}, @{$device->{'methods'}}) {
+                    return $ACTIONS{$self->radius_mfa_method}->($self,$device->{'device'},$username,$self->radius_mfa_method);
+                } else {
+                    $logger->info("Unsuported method on device ".$device->{'name'});
```
Comment (jrouzierinverse) [inline lib/pf/mfa/Akamai.pm]: Typo ```Unsuported -> Unsupported```

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

**Comment by jrouzierinverse** [comment]
Everything looks good.
@julsemaan if you agree I will merge it.

**Comment by julsemaan** [comment]
> Everything looks good. @julsemaan if you agree I will merge it.

Yep, I agree, I had my 'Approved' status for my review on this one

**Pull request merged**